### PR TITLE
feat: add interfaces

### PR DESCRIPTION
**Pull request by redeboer**
Closes https://github.com/ComPWA/tensorwaves/issues/18

Defines the basic interfaces for the interaction of the main modules of
tensorwaves:

* Intensity/Function
* Estimator
* Data
* Optimizer

Note that the interfaces are inspired by ComPWA.
https://github.com/ComPWA/ComPWA

### CODE REVIEW

File: tensorwaves/interfaces.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        pass
+
+
+class Kinematics(ABC):
+    @abstractmethod
+    def convert(self, events: dict) -> dict:
+        pass
+
+    @abstractmethod
+    def reduce_to_phase_space(self, events: dict) -> dict:
```
Comment (redeboer) [inline tensorwaves/interfaces.py:66]: Is this necessary? Or does a `is_within_phase_space` suffice?

File: tensorwaves/interfaces.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
+"""Defines top level interfaces of tensorwaves.
+
+"""
+
+from abc import ABC, abstractmethod
+
+
+class Function(ABC):
```
Comment (redeboer) [inline tensorwaves/interfaces.py:8]: Sufficient for treating intensities? Or should there be a derived intensity class or so?

File: tensorwaves/interfaces.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        """
+
+    @parameters.setter
+    @abstractmethod
+    def parameters(self, new_parameters: dict):
+        """
+        """
+
+
+class Estimator(ABC):
```
Comment (redeboer) [inline tensorwaves/interfaces.py:44]: What is an 'estimator' precisely? I can't formulate it in another way than something like 'estimate of the discrepancy between model and data set'. That suggests there should be some link to both a model and a data set.

File: tensorwaves/interfaces.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        pass
+
+
+class Optimizer(ABC):
+    @abstractmethod
+    def optimize(self, estimator: Estimator, initial_parameters: dict) -> dict:
+        pass
+
+
+class PhaseSpaceGenerator(ABC):
```
Comment (redeboer) [inline tensorwaves/interfaces.py:81]: Is a phase space generator independent of a uniform generator?

File: tensorwaves/interfaces.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    @abstractmethod
+    def generate(self, size: int, random_generator) -> dict:
+        pass
+
+
+class UniformRealNumberGenerator(ABC):
```
Comment (redeboer) [inline tensorwaves/interfaces.py:87]: Does a real number generator have to be uniform?

File: tensorwaves/interfaces.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    def reduce_to_phase_space(self, events: dict) -> dict:
+        pass
+
+    @property
+    @abstractmethod
+    def phase_space_volume(self) -> float:
+        pass
+
+
+class Optimizer(ABC):
```
Comment (redeboer) [inline tensorwaves/interfaces.py:75]: An optimizer has to optimize a model. Should some reference to a model be implemented in this abstract interface?

File: tensorwaves/interfaces.py (hunk 1 above)
Comment (spflueger) [reply tensorwaves/interfaces.py:66]: I like `is_within_phase_space` better

File: tensorwaves/interfaces.py (hunk 2 above)
Comment (spflueger) [reply tensorwaves/interfaces.py:8]: yeah `Intensity` would be just a special type of `Function` (regarding the return value)

File: tensorwaves/interfaces.py (hunk 3 above)
Comment (spflueger) [reply tensorwaves/interfaces.py:44]: Yeah its exactly that. But in an interface should not define any state, that why it is this way.
If data and model would be added to `__call__`  then we static property of the comparison is lost. So I don't know any better way

Comment (redeboer) [reply tensorwaves/interfaces.py:44]: What is the static property of the comparison?

File: tensorwaves/interfaces.py (hunk 4 above)
Comment (spflueger) [reply tensorwaves/interfaces.py:81]: Yeah, the idea here is that we want to have full reproducability or deterministic behavior. So if you chose a specific phase space generator and seed, the output should always be the same. Since you need the random number generation in the generation of phase space events and in possible hit & miss / accept and reject, it was best to hand a random generator to the phase space generator. That avoids mixups (having two random generators and maintaining their state) and also gives flexibility.

File: tensorwaves/interfaces.py (hunk 5 above)
Comment (spflueger) [reply tensorwaves/interfaces.py:87]: We only need uniform real number generation. I guess if you need something more special you can also define custom distributions that use that uniform generation as input.

File: tensorwaves/interfaces.py (hunk 6 above)
Comment (spflueger) [reply tensorwaves/interfaces.py:75]: The optimizer get an Estimator and a set of parameters. He has to tweak the parameters so that the Estimator return value becomes as small as possible

File: tensorwaves/interfaces.py (hunk 1 above)
Comment (redeboer) [reply tensorwaves/interfaces.py:66]: So the following?
```python
def is_within_phase_space(self, event) -> bool:
    """Check whether event lies within phase space."""
```
Or maybe this?
```python
def is_within_phase_space(self, events: dict) -> list:
    """Check which events lie within phase space."""
```

File: tensorwaves/interfaces.py (hunk 6 above)
Comment (redeboer) [reply tensorwaves/interfaces.py:75]: Then I suppose this comment resolves this https://github.com/ComPWA/tensorwaves/pull/30/files/a2d859cf1dfd332e6e18f4f97a2bad5faea54911#r413062124

File: tensorwaves/interfaces.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    @abstractmethod
+    def optimize(self, estimator: Estimator, initial_parameters: dict) -> dict:
+        pass
+
+
+class PhaseSpaceGenerator(ABC):
+    @abstractmethod
+    def generate(self, size: int, random_generator) -> dict:
```
Comment (redeboer) [inline tensorwaves/interfaces.py:83]: Probably
```python
def generate(self, size: int, random_generator: UniformRealNumberGenerator) -> dict:
```
(note the `UniformRealNumberGenerator`)

File: tensorwaves/interfaces.py (hunk 4 above)
Comment (redeboer) [reply tensorwaves/interfaces.py:81]: See https://github.com/ComPWA/tensorwaves/pull/30/files/a2d859cf1dfd332e6e18f4f97a2bad5faea54911#r413077761
Or should the `random_generator` in `generate` be a `seed` (`int` or `float`)?

File: tensorwaves/interfaces.py (hunk 3 above)
Comment (spflueger) [reply tensorwaves/interfaces.py:44]: That the model and the data stay fixed for one instance of an estimator. Only parameters of the model (which are also the parameters of the estimator) are the things that will/can change

File: tensorwaves/interfaces.py (hunk 1 above)
Comment (spflueger) [reply tensorwaves/interfaces.py:66]: Definitely the second one. Because the first thing we had in the beginning and this event based way of processing did not play together well with the vectorized forms in the rest of the code

File: tensorwaves/interfaces.py (hunk 7 above)
Comment (spflueger) [reply tensorwaves/interfaces.py:83]: yeah sorry, thats what I had before already. (that got lost in my screw up before)

File: tensorwaves/interfaces.py (hunk 1 above)
Comment (redeboer) [reply tensorwaves/interfaces.py]: https://github.com/ComPWA/tensorwaves/pull/30/commits/d55d6b53324828ea50f359b2dae0870fbdae42ea

File: tensorwaves/interfaces.py (hunk 7 above)
Comment (redeboer) [reply tensorwaves/interfaces.py]: https://github.com/ComPWA/tensorwaves/pull/30/commits/abe46ecb56db3f9abca23211a3cee3eaa9c65bce

File: tensorwaves/interfaces.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    def optimize(self, estimator: Estimator, initial_parameters: dict) -> dict:
+        """Execute optimization."""
+
+
+class UniformRealNumberGenerator(ABC):
+    """Abstract class for generating uniform real numbers."""
+
+    @abstractmethod
+    def __call__(self, size: int,
+                 min_value=0.0, max_value=1.0) -> float or list:
```
Comment (redeboer) [inline tensorwaves/interfaces.py:91]: Currently, the types are only as hints to the user and for the API (https://github.com/ComPWA/tensorwaves/pull/27) — there are no checks on these. For better typing, it is better to use the [`typing`](https://docs.python.org/3/library/typing.html) module (from Python 3.5, so should be fine) along with [`mypy`](http://mypy-lang.org/). This can also be added as a test to Travis CI.

Comment (spflueger) [reply tensorwaves/interfaces.py:91]: https://stackoverflow.com/questions/33945261/how-to-specify-multiple-return-types-using-type-hints

Btw you are right, this `or` is not valid. I guess its fine we add this in PR #33

File: .pylintrc [hunk 9]
```diff
@@ -6,4 +6,4 @@
 init-import=yes
 
 [MESSAGES CONTROL]
-disable=RST203, RST301
+disable=RST203, RST301, E0601
```
Comment (spflueger) [inline .pylintrc:9]: Why did you need to add E0601 (used-before-assignment). Where does this happen here?

Comment (redeboer) [reply .pylintrc:9]: In https://github.com/ComPWA/tensorwaves/pull/30/commits/abe46ecb56db3f9abca23211a3cee3eaa9c65bce, `UniformRealNumberGenerator` was defined before `PhaseSpaceGenerator` which generated the `E0610` error. That order changed in https://github.com/ComPWA/tensorwaves/pull/30/commits/7693bca69d6d33a164a4308b03dc567d8ada9cb9, but I did not remove the pylint flag.

Should the E0610 ignore flag be removed?

Comment (spflueger) [reply .pylintrc:9]: yeah

Comment (redeboer) [reply .pylintrc]: https://github.com/ComPWA/tensorwaves/pull/30/commits/daa36ebc6b29118bc82bfccc68d7f6d9dc4c8a08

File: tensorwaves/interfaces.py (hunk 8 above)
Comment (redeboer) [reply tensorwaves/interfaces.py]: https://github.com/ComPWA/tensorwaves/pull/30/commits/541a23598565c09ff105474d92cab4910bf4a9e2

File: tox.ini [hunk 10]
```diff
@@ -23,4 +23,5 @@ exclude =
     __pycache__
 ignore =
     W503
+    RST304 # Unknown interpreted text role
```
Comment (redeboer) [inline tox.ini:26]: Maybe this ignore should actually not be added. The warning it resulted in may have been correct and would have been picked out more accurately by https://github.com/ComPWA/tensorwaves/pull/27.

Comment (spflueger) [reply tox.ini:26]: yeah, better we add a custom defined role instead of ignoring all

Comment (redeboer) [reply tox.ini]: https://github.com/ComPWA/tensorwaves/pull/30/commits/063b260e45411f101b567406bbe92f07e5ca3bd1
